Documentation/ceph-best-practices.md

```diff
@@ -0,0 +1,516 @@
+  OSD maintenance and no need to delete Deployments for OSDs that have been kicked out. Rook will
+  automatically clean up the cluster by removing OSD Pods if the OSDs are no longer holding Ceph
+  data. However, keep in mind that this can reduce the visibility of failures from Kubernetes Pod
+  and Pod Controller views. You can optionally set `removeOSDSIfOutAndSafeToRemove` to `false` if
+  need be, such as if a Kubernetes administrator wants to see disk failures via a Pod overview.
+* Configure Ceph using the central configuration database when possible. This allows for more
+  runtime configuration flexibility. Do this using `ceph config set` commands from Rook's toolbox.
+  Only use Rook’s provided ceph.conf override ConfigMap when it is required.
+
+
+## Limiting Ceph to Specific Nodes
+
+One of the more common setups you may want for your Rook-Ceph cluster is to limit Ceph to a specific
+set of nodes. Even for co-located use cases, you could have valid reasons why you must not (or do
+not want to) use some nodes for storage. This is applicable for both co-located and disaggregated
+use cases. To limit Ceph to specific nodes, we can label Kubernetes nodes and configure Rook to have
+Affinity (as a hard preference) for our label.
+
+For our example, let's label the desired storage nodes with `storage-node=true`. To run Rook and
+Ceph daemons on labeled nodes, we will configure Affinities for this label in both the Rook
+Operator manifest (`operator.yaml`) and the Ceph cluster manifest (`cluster.yaml`).
+
+For Rook daemons and the CSI driver daemons, adjust the Operator manifest. Note that the CSI
+Provisioner is best started on the same nodes as other Ceph daemons.
+
+**`operator.yaml`:**
+```yaml
+CSI_PROVISIONER_NODE_AFFINITY: “storage-node=true”
+AGENT_NODE_AFFINITY: “storage-node=true”
+DISCOVER_AGENT_NODE_AFFINITY: “storage-node=true”
+```
+
+For `all` Ceph daemons, we must add Affinity for storage nodes in `cluster.yaml`.
+
+**`cluster.yaml`:**
+```yaml
+placement:
+  all:
+    nodeAffinity:
+      requiredDuringSchedulingIgnoredDuringExecution:
+        nodeSelectorTerms:
+          - matchExpressions:
+              - key: storage-node
+                operator: In
+                values:
+                  - "true"
+```
+
+## Segregating Ceph from User Applications
+
+You could also have reason to totally separate Rook and Ceph nodes from application nodes. This
+falls under the disaggregated use-case, and it is a more traditional way to deploy storage. In this
+case, we still need to follow the best practices described in the section above to
+[Limit Ceph to Specific Nodes](#limiting-ceph-to-specific-nodes), and we also need some additional
+settings.
+
+To segregate Ceph from user applications, we will also label all non-storage nodes with
+`storage-node=false`. The CSI plugin pods must run where user application Pods run and not where
```

Comment:
Why is a `storage-node=false` label needed on the non-Ceph nodes? The tolerations in the cluster.yaml don't apply to the CSI driver, so the CSI driver will act just like the application pods and avoid nodes with the taint.

Reply:
We need to schedule the drivers on nodes where Ceph pods are **not** running, but we don't support anti-affinities in Rook. So the only thing we can do is create positive affinity for the inverse label (`storage-node=false`). If we don't do this, we are unnecessarily running CSI pods on Ceph nodes, and that wastes resources.

Reply:
How would the CSI driver be different from application pods? Since the nodes running Ceph daemons are tainted, the taint will cause the CSI driver to act like the applications and avoid that node. You just wouldn't add the toleration to the csi driver.